src/main/scala/com/avast/clients/stor/FileCopier.scala

```diff
@@ -0,0 +1,71 @@
+package com.avast.clients.stor
+
+import java.io.{InputStream, OutputStream}
+import java.security.MessageDigest
+
+import org.apache.commons.codec.binary.Hex
+import org.apache.commons.io.IOUtils
+
+class FileCopier {
+
+  private val lock = new Object
+
+  private val digest: MessageDigest = MessageDigest.getInstance("SHA-256")
```

Comment:
Please use this code instead - the you don't have to use `lock.synchronized`:
```scala
private val sha256providers = new ThreadLocal[MessageDigest]() {
    override def initialValue(): MessageDigest = MessageDigest.getInstance("SHA-256")
  }
```
Usage: `sha256providers.get().digest(...)`

Reply:
Ah, you are using still a new instance of this class. So I would prefer to extract the `sha256providers` definition to a _private_ `object`, so we don't have to create a new instance of `MessageDigest` again and again (it has some overhead).

Reply:
Done